pkg/api/api.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/prometheus/common/model"
+)
+
+const (
+	statusAPIError = 422
+
+	// Possible values for ErrorType.
+	ErrBadResponse promapiv1.ErrorType = "bad_response"
+	ErrServer      promapiv1.ErrorType = "server_error"
+	ErrClient      promapiv1.ErrorType = "client_error"
+
+	epQueryRange = "/api/v1/query_range"
+)
+
+func errorTypeAndMsgFor(resp *http.Response) (promapiv1.ErrorType, string) {
+	switch resp.StatusCode / 100 {
+	case 4:
+		return ErrClient, fmt.Sprintf("client error: %d", resp.StatusCode)
+	case 5:
+		return ErrServer, fmt.Sprintf("server error: %d", resp.StatusCode)
+	}
+	return ErrBadResponse, fmt.Sprintf("bad response code %d", resp.StatusCode)
+}
+
+func apiError(code int) bool {
+	// These are the codes that Prometheus sends when it returns an error.
+	return code == statusAPIError || code == http.StatusBadRequest
+}
+
+// queryResult contains result data for a query.
+type queryResult struct {
+	Type   model.ValueType `json:"resultType"`
+	Result interface{}     `json:"result"`
+
+	// The decoded value.
+	v model.Value
+}
+
+type apiResponse struct {
+	Status    string              `json:"status"`
+	Data      json.RawMessage     `json:"data"`
+	ErrorType promapiv1.ErrorType `json:"errorType"`
+	Error     string              `json:"error"`
+	Warnings  []string            `json:"warnings,omitempty"`
+}
+
+func do(ctx context.Context, client promapi.Client, req *http.Request) (*http.Response, []byte, promapiv1.Warnings, error) {
+	resp, body, err := client.Do(ctx, req)
+	if err != nil {
+		return resp, body, nil, err
+	}
+
+	code := resp.StatusCode
+
+	if code/100 != 2 && !apiError(code) {
+		errorType, errorMsg := errorTypeAndMsgFor(resp)
+		return resp, body, nil, &promapiv1.Error{
+			Type:   errorType,
+			Msg:    errorMsg,
+			Detail: string(body),
+		}
+	}
+
+	var result apiResponse
+
+	if http.StatusNoContent != code {
+		if jsonErr := json.Unmarshal(body, &result); jsonErr != nil {
+			return resp, body, nil, &promapiv1.Error{
+				Type: ErrBadResponse,
+				Msg:  jsonErr.Error(),
+			}
+		}
+	}
+
+	if apiError(code) && result.Status == "success" {
+		err = &promapiv1.Error{
+			Type: ErrBadResponse,
+			Msg:  "inconsistent body for response code",
+		}
+	}
+
+	if result.Status == "error" {
+		err = &promapiv1.Error{
+			Type: result.ErrorType,
+			Msg:  result.Error,
+		}
+	}
+
+	return resp, result.Data, result.Warnings, err
+}
+
+// doGetFallback will attempt to do the request as-is, and on a 405 it will fallback to a GET request.
+func doGetFallback(
+	ctx context.Context,
+	client promapi.Client,
+	u *url.URL,
+	args url.Values,
+	cache bool,
+) (*http.Response, []byte, promapiv1.Warnings, error) {
+	req, err := http.NewRequest(http.MethodPost, u.String(), strings.NewReader(args.Encode()))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
+	if !cache {
+		req.Header.Set("Cache-Control", "no-store")
```

Comment:
I copied the client_golang code to add support for setting the cache-control header if we don't want the result to be cached.

Reply:
hm... so this is great to not cache things, but it does not stop using cache on entry (hit), no? (: I think we want to omit caching totally :thinking:

Reply:
It omits all things. Both cache results and read from cache

Reply:
https://github.com/cortexproject/cortex/blob/master/pkg/querier/queryrange/results_cache.go#L187-L189
If this header is set, it goes to the next middleware. So it omits all cache things.

Reply:
All good!